### PR TITLE
Fearute_Logic_Register

### PR DESCRIPTION
**Pull request by tantuan127**

### CODE REVIEW

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterPresenter.java [hunk 1]
```diff
@@ -20,4 +27,32 @@ public void onStart() {
...
+                    public void call(Throwable throwable) {
+                        mViewModel.onRegisterError();
+                    }
+                });
+    }
+
+    @Override
+    public boolean validateDataInput(String userName, String passWord, String confirmPassword, String firstName,
+            String lastName, String address, String role, String department) {
+        if (TextUtils.isEmpty(userName) && TextUtils.isEmpty(passWord) && TextUtils.isEmpty(confirmPassword) && TextUtils.isEmpty(firstName) && TextUtils.isEmpty(lastName) && TextUtils.isEmpty(lastName) && TextUtils.isEmpty(address) && TextUtils.isEmpty(role) && TextUtils.isEmpty(department)) {
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterPresenter.java]: if (true) {
return false
}
return true

replace by
return TextUtils.isEmpty(userName) && TextUtils.isEmpty(passWord)...

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java [hunk 2]
```diff
@@ -1,14 +1,32 @@
...
 /**
  * Exposes the data to be used in the Register screen.
  */
 
-public class RegisterViewModel implements RegisterContract.ViewModel {
+public class RegisterViewModel extends BaseObservable implements RegisterContract.ViewModel {
 
+    private Context mContext;
     private RegisterContract.Presenter mPresenter;
+    private String mUsername;
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java]: Sao e ko sử dụng model User nhỉ? mà phải khai báo nhiều instance variable thế này

File: app/src/main/java/com/framgia/fdms/screen/login/LoginPresenter.java [hunk 3]
```diff
@@ -49,7 +50,7 @@ public void call(Throwable throwable) {
 
     @Override
     public boolean validateDataInput(String username, String password) {
-        if (TextUtils.isEmpty(username) && TextUtils.isEmpty(password)) {
+        if (TextUtils.isEmpty(username) || TextUtils.isEmpty(password)) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/login/LoginPresenter.java]: (y)

Comment (DoanVanToan) [inline app/src/main/java/com/framgia/fdms/screen/login/LoginPresenter.java]: return !(TextUtils.isEmpty(username) || TextUtils.isEmpty(password))

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterContract.java [hunk 4]
```diff
@@ -11,11 +11,32 @@
...
+        void onInputRoleError();
+
+        void onInputDepartmentError();
     }
 
-    /**
-     * Presenter.
-     */
     interface Presenter extends BasePresenter {
+        void register(String userName, String passWord, String confirmPassword, String firstName,
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterContract.java]: should replace parram to
register(User user)
validateDataInput (User user)

**Comment by daolq3012** [comment]
@tantuan127 Please change commit name

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterContract.java (hunk 4 above)
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterContract.java]: Please create RegisterRequest include all data send to server and it is param for this method :)

Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterContract.java]: 1 method maximum 5 params

File: app/src/main/java/com/framgia/fdms/data/model/User.java [hunk 5]
```diff
@@ -7,23 +7,28 @@
...
+    public String mConfirmPassword;
     public String mFirstName;
     public String mLastName;
     public String mAddress;
     public String mRole;
     public String mDepartment;
 
-    public User(String userName, String passWord, String firstName, String lastName, String address,
-            String role, String department) {
+    public User(String userName, String passWord, String confirmPassword, String firstName,
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/data/model/User.java]: ``User`` not contain password and confirm password, please remove that params

File: app/src/main/java/com/framgia/fdms/data/source/remote/UserRemoteDataSource.java [hunk 6]
```diff
@@ -19,7 +20,13 @@ public UserRemoteDataSource(FDMSApi FDMSApi) {
...
         // TODO: 4/4/17 replace by call API later
-        User user = new User("tuan", "123", "", "", "", "", "");
+        User user = new User("tuan", "123", "", "", "", "", "", "");
         return Observable.just(user);
     }
+
+    @Override
+    public Observable<RegisterRequest> register(RegisterRequest request) {
+        // TODO: 4/4/17 replace by call API later
+        return Observable.just(request);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/data/source/remote/UserRemoteDataSource.java]: when register success, it does not return a request, usually return an object ``user`` or just a simple response
Please change Object return

File: app/src/main/java/com/framgia/fdms/screen/login/LoginPresenter.java [hunk 7]
```diff
@@ -49,9 +50,15 @@ public void call(Throwable throwable) {
 
     @Override
     public boolean validateDataInput(String username, String password) {
-        if (TextUtils.isEmpty(username) && TextUtils.isEmpty(password)) {
-            return false;
+        boolean isValid = true;
+        if (TextUtils.isEmpty(username)) {
+            isValid = false;
+            mView.onInputUserNameError();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/login/LoginPresenter.java]: greate!

File: app/src/main/java/com/framgia/fdms/data/source/api/request/RegisterRequest.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
+package com.framgia.fdms.data.source.api.request;
+
+/**
+ * Created by levutantuan on 4/5/17.
+ */
+
+public class RegisterRequest {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/data/source/api/request/RegisterRequest.java]: Request should extends BaseRequest

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterActivity.java [hunk 9]
```diff
@@ -16,12 +20,14 @@
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        RegisterRequest request = new RegisterRequest();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterActivity.java]: Please init ``RegisterRequest`` inside ``RegisterViewModel``

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterPresenter.java [hunk 10]
```diff
@@ -20,4 +36,64 @@ public void onStart() {
...
+                    public void call(User request) {
+                        mViewModel.onRegisterSuccess();
+                    }
+                }, new Action1<Throwable>() {
+                    @Override
+                    public void call(Throwable throwable) {
+                        mViewModel.onRegisterError();
+                    }
+                });
+        mCompositeSubscription.add(subscription);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterPresenter.java]: add ``mCompositeSubscription.clear()`` in ``onStop()``

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java [hunk 11]
```diff
@@ -1,14 +1,43 @@
...
+    private String mRole;
+    private String mDepartment;
+    private String mUsernameError;
+    private String mPasswordError;
+    private String mConfirmPasswordError;
+    private String mFirstnameError;
+    private String mLastnameError;
+    private String mAddressError;
+    private String mRoleError;
+    private String mDepartmentError;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java]: remove all variables, It unnecessary

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java [hunk 12]
```diff
@@ -28,5 +57,187 @@ public void setPresenter(RegisterContract.Presenter presenter) {
 
     public void onCreateAccountClick() {
         //TODO Create New User
+        RegisterRequest request =
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java]: remove, Let use ``mRequest``

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java [hunk 13]
```diff
@@ -28,5 +57,187 @@ public void setPresenter(RegisterContract.Presenter presenter) {
...
+
+    @Override
+    public void onInputDepartmentError() {
+        mDepartmentError = mContext.getString(R.string.msg_error_user_name);
+        notifyPropertyChanged(com.framgia.fdms.BR.departmentError);
+    }
+
+    @Bindable
+    public String getUserName() {
+        return mUsername;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java]: change to
```java
return mRequest.getUserName();
```

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java [hunk 14]
```diff
@@ -28,5 +57,187 @@ public void setPresenter(RegisterContract.Presenter presenter) {
...
+    public void setDepartment(String department) {
+        mDepartment = department;
+    }
+
+    public String getUsername() {
+        return mUsername;
+    }
+
+    public void setUsername(String username) {
+        mUsername = username;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterViewModel.java]: change to
```java
mRequest.setUserName(username);
```

File: app/src/main/java/com/framgia/fdms/screen/register/RegisterPresenter.java [hunk 15]
```diff
@@ -1,23 +1,100 @@
...
+
 /**
  * Listens to user actions from the UI ({@link RegisterActivity}), retrieves the data and updates
  * the UI as required.
  */
 final class RegisterPresenter implements RegisterContract.Presenter {
-    private static final String TAG = RegisterPresenter.class.getName();
 
     private final RegisterContract.ViewModel mViewModel;
+    private CompositeSubscription mCompositeSubscription = new CompositeSubscription();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fdms/screen/register/RegisterPresenter.java]: should init in contructor method
Please update!

**Comment by daolq3012** [comment]
@tantuan127 Great!
Change commit name (tat)

**Comment by daolq3012** [comment]
OK!

**Pull request merged**